File: progs/javascript/scope.js
//Scope: unscoped
let counter = 0;

async function print(msg, signal) {
  console.log(msg);
}

async function asyncInt(signal) {
  let inner = async () => {
    let me = counter++;
    let until = Math.floor(Math.random() * 50);
    for (let i = 0; i < until; i++)
      await print(`[${me}] waiting`, signal);
    await print(`[${me}] done`);
    return until;
  };
  inner();

  return 0;
}

async function main() {
  let a = asyncInt();
  let b = asyncInt();
  console.log("working");
  let c = await a + await b;
  await print(c);
}

main().then(() => console.log("exiting"))

File: progs/python/aio/scope.py
#Scope: program
import asyncio
from random import randint

counter = 0
background = []

async def write(msg):
    print(msg)
    await asyncio.sleep(0.5)

async def async_int():
    global background
    async def inner():
        global counter
        me = counter
        counter += 1
        until = randint(0, 50)
        for _ in range(until):
            await write(f"{me} waiting")
        print(f"{me} done")
        return until
    background.append(asyncio.create_task(inner()))
    return 0

async def main():
    a = async_int()
    b = async_int()
    print("working")
    c = await a + await b
    await write(c)

asyncio.run(main())
